### PR TITLE
[SPARK-23451][ML] Deprecate KMeans.computeCost

### PR DESCRIPTION
**Pull request by mgaido91**
## What changes were proposed in this pull request?

Deprecate `KMeans.computeCost` which was introduced as a temp fix and now it is not needed anymore, since we introduced `ClusteringEvaluator`.

## How was this patch tested?

manual test (deprecation warning displayed)
Scala
```
...
scala> model.computeCost(dataset)
warning: there was one deprecation warning; re-run with -deprecation for details
res1: Double = 0.0
```

Python
```
>>> import warnings
>>> warnings.simplefilter('always', DeprecationWarning)
...
>>> model.computeCost(df)
/Users/mgaido/apache/spark/python/pyspark/ml/clustering.py:330: DeprecationWarning: Deprecated in 2.4.0. It will be removed in 3.0.0. Use ClusteringEvaluator instead.
  " instead.", DeprecationWarning)
```

### CODE REVIEW

**Comment by MLnick** [comment]
Just want to check - does `computeCost` do the same thing as the silhouette metric?

**Comment by mgaido91** [comment]
thanks for taking a look at this @MLnick. No, it doesn't, in the sense that it returns a different result: this is the sum of the squared euclidean distance between a point and the centroid of the cluster it is assigned to, while the silhouette metric is the average of the silhouette coefficient. So they are completely different formulas.

The semantic is a bit different too. Silhouette measures both cohesion and separation of the clusters, while `computeCost` as it is measures only cohesion.

Nonetheless, of course both them can be used to evaluate the result of a clustering algorithm, even though the silhouette is much better for this purpose.

**Comment by MLnick** [comment]
Right - so while it’s perhaps a lower quality metric it is different. So I
wonder if deprecation is the right approach (vs say putting the within
cluster sum squares into ClusteringEvaluator).

**Comment by MLnick** [comment]
Sorry I mean putting the metric in evaluator and then also deprecating
computCost

**Comment by mgaido91** [comment]
yes, I agree with you @MLnick. I'd like to ping also @jkbradley and @hhbyyh who drove the PR which introduce `computeCost`.

In order to move the same metric to `ClusteringEvaluator`, I think it is very useful what is done in this ongoing PR #20600, where `DistanceMeasure` is moved to its own file and the `cost` method is added to it. So maybe it can be added there after that PR. What do you think?

PS if you agree, may you please then help reviewing that PR?

Thanks.

**Comment by mgaido91** [comment]
sorry @MLnick, what do you think about my previous comment? Any thoughts? Thanks.

**Comment by mgaido91** [comment]
@MLnick I checked and adding the `computeCosts` to `ClusteringEvaluator` has a small drawback: we have to compute the centers for each cluster and then we can compute the costs, which involved two passes on the dataset.

Since this and that this evaluation looks not very useful in practice, is it worth according to you to add it nonetheless?

Thanks.

**Comment by mgaido91** [comment]
kindly ping @MLnick  @jkbradley @hhbyyh

**Comment by mgaido91** [comment]
kindly ping @MLnick @jkbradley @hhbyyh

**Comment by holdenk** [comment]
So when you say "second pass over the data" - from looking at this it seems like it would could do this with just a second map to look up the predictions in the already computed cluster centers, not a stage boundary, so that probably wouldn't be all that expensive given how Spark does pipe-lining unless I'm mussing something.

This would mean that we'd have to have people set the cluster centers from their model when they wanted to do that evaluation type but given that the evaluate wouldn't be able to recover the cluster centers from a test that differed from the training set I think that would be reasonable.

That being said its been awhile since I've looked at the evaluator code so I could be coming out of left field.

**Comment by mgaido91** [comment]
@holdenk I am not sure I got 100% what you meant, so I'll try to answer but let me know if I missed something.

The problem of doing 2 passes is related to cluster centers. The API of `ClusteringEvaluator` (as of any `Evaluator`) is very simple: it is has a method which gets a `Dataset` and returns a value. So, unlike the method here - which is part of the `KMeansModel` and it can get the cluster centers from it -, there is no clue about the cluster centers: computing them is easy but it requires a pass on the dataset (this is the extra pass I mentioned).

An alternative to this is adding a `setClusterCenters` method on the `ClusteringEvaluator`, but I am not sure whether this is worth since they are needed only for this metric, while for the others so far (the Silhouette measure) they are useless. Moreover, this metric was introduced explicitly as a temp fix because we were missing any other (better) evaluation metric and it was supposed to be dismissed once a better evaluation metric would have been introduced (please see the related JIRA and PR). So I am not sure that introducing a new method specifically for this metric is a good idea.

What do you think? Were you suggesting this second option?

**Review by holdenk — COMMENTED** [review]
Left some draft comments, lets see what @MLnick thinks of this. If he doesn't have time that's ok but I'd like his input.

I think after seeing this with the new param it maybe makes sense to not require the param (sorry) and do the double computation when the param isn't set so that the distance measure is more like the others but we also don't introduce a slow down for folks moving from the deprecated code path. What do you think?

File: python/pyspark/ml/clustering.py [hunk 1]
```diff
@@ -322,7 +323,11 @@ def computeCost(self, dataset):
         """
         Return the K-means cost (sum of squared distances of points to their nearest center)
         for this model on the given data.
+
+        ..note:: Deprecated in 2.4.0. It will be removed in 3.0.0. Use ClusteringEvaluator instead.
         """
+        warnings.warn("Deprecated in 2.4.0. It will be removed in 3.0.0. Use ClusteringEvaluator"
```
[the remark this answers is not in the document]
Comment (mgaido91) [reply python/pyspark/ml/clustering.py]: yes, or I can also update it here once we establish for sure what the new API has to look like, as you prefer.

File: mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala [hunk 2]
```diff
@@ -64,12 +65,12 @@ class ClusteringEvaluator @Since("2.3.0") (@Since("2.3.0") override val uid: Str
 
   /**
    * param for metric name in evaluation
-   * (supports `"silhouette"` (default))
+   * (supports `"silhouette"` (default), `"kmeansCost"`)
```
[the remark this answers is not in the document]
Comment (mgaido91) [reply mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala]: but does it make sense to introduce something which is already considered legacy when introduced?

I think this brigs up again the question: shall we maintain a metric which was introduced only temporary as a fallback due to the lack of better metrics?

**Comment by mgaido91** [comment]
@holdenk I am not sure about requiring or not cluster centers for this metric. On one side, since the `ClusteringEvaluator` should be a general interface for all clustering models and some of them don't provide cluster centers, it may be a good idea to compute them if necessary. On the other, does this metric make sense for any model other than KMeans? And computing the centers of the test dataset would lead to different results than the old API we are replacing. So I am not sure it is the right thing to do.

Honestly, the more we go on the more my feeling is that we don't really need to move that metric here. We can just deprecate it saying that there are better metrics for evaluating a clustering available in the `ClusteringEvaluator` (namely the silhouette). In these way people can move away from using this metric.

Moreover, sklearn - which is one of the most widespread tool - doesn't offer the ability of computing such a cost (http://scikit-learn.org/stable/modules/clustering.html#clustering-performance-evaluation). The only thing sklearn offers is what it calls `inertia` (https://github.com/scikit-learn/scikit-learn/blob/a24c8b46/sklearn/cluster/k_means_.py#L265), ie. the cost computed on the training set.

So, I think the best option would be to follow what sklearn does:

 1 - Introducing in the `KMeansSummary` (or `KMeansModel` if you prefer) the cost attribute on the training set
 2 - deprecate this method redirecting to `ClusteringEvaluator` for better metrics and/or to the cost attribute introduced

What do you think?

**Comment by mgaido91** [comment]
kindly ping @holdenk @MLnick

**Comment by mgaido91** [comment]
kindly ping @holdenk @MLnick

File: mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (holdenk) [reply mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala]: Generally speaking I think it would make sense to maintain the fall-back metric until at least Spark 3.0 at which point I think it would make sense to ask on the user and dev lists and see if anyone is hard dependencies on it or if it is safe to remove.

**Comment by holdenk** [comment]
cc @sethah

File: mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala (hunk 2 above)
[the remark this answers is not in the document]
Comment (mgaido91) [reply mllib/src/main/scala/org/apache/spark/ml/evaluation/ClusteringEvaluator.scala]: ok, I agree. Let's go on this way then, thanks.

**Comment by holdenk** [comment]
ping @sethah :)

**Comment by holdenk** [comment]
ping @sethah?

**Comment by sethah** [comment]
+1 for @mgaido91's plan

**Comment by mgaido91** [comment]
@holdenk @sethah any more comments on this? As 2.4 release is approaching I think it would be great to have this in, so we can remove the deprecated APIs in 3.0... Thanks.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeansModel.scala [hunk 3]
```diff
@@ -37,7 +37,7 @@ import org.apache.spark.sql.{Row, SparkSession}
  */
 @Since("0.8.0")
 class KMeansModel @Since("2.4.0") (@Since("1.0.0") val clusterCenters: Array[Vector],
-  @Since("2.4.0") val distanceMeasure: String)
+  @Since("2.4.0") val distanceMeasure: String, @Since("2.4.0") val trainingCost: Double)
```
Comment (holdenk) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeansModel.scala]: Since we changed the constructor here, and since it is not private, we should provide a similar (and deprecated) constructor without training cost which calls this with the default value.

**Review by holdenk — CHANGES_REQUESTED** [review]
Love this different approach, I agree we should get this in before 2.4 closes so we can remove in 3 line. One comment though for API compatibility with the constructor. Really hope we can merge this soon :)

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeansModel.scala (hunk 3 above)
Comment (mgaido91) [reply mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeansModel.scala]: This constructor was introduced by a previous PR for 2.4, so this was never out (therefore I think we don't need to keep it).

Comment (holdenk) [reply mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeansModel.scala]: That's awesome, lets try and get that fixed up before 2.4 goes so we don't have to any workarounds :)

**Comment by holdenk** [comment]
After I merged your other PR this has some minor conflicts so needs an update, but I'd be happy to try and get this in end of next week during my next review session :)

**Review by holdenk — APPROVED** [review]
LGTM thanks for getting this ready for removal in 3 :)

**Comment by holdenk** [comment]
Merged to master :) Thank you :)
